File: src/familiar_connect/sleep/apply.py
"""Apply a validated :class:`ConsolidationPlan` to the DB.

Separate from planning so a dry run can compute the plan without
touching the DB. Apply is supersede-only: retires drop facts with no
replacement, rewrites append one consolidated fact then supersede the
old ones by it. The sleep watermark advances once, to the window's
high-water marks, regardless of how many actions ran.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from familiar_connect.history.store import FactDraft, FactSubject
from familiar_connect.identity import is_self_key

if TYPE_CHECKING:
    from familiar_connect.history.async_store import AsyncHistoryStore
    from familiar_connect.sleep.consolidation import ConsolidationPlan, RewriteAction

_logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class ApplyReport:
    """Outcome of applying a plan.

    ``skipped`` records actions a concurrent writer invalidated between
    plan and apply — each ``(kind, fact_id, reason)``. Per-action
    ValueError handling makes apply non-fatal and idempotent: a partial
    or already-applied plan re-runs cleanly instead of crashing.
    """

    retired_fact_ids: tuple[int, ...]
    rewritten: tuple[tuple[tuple[int, ...], int], ...]  # (old_ids, new_id)
    watermark: tuple[int, int]  # (last_fact_id, last_turn_id)
    skipped: tuple[tuple[str, int, str], ...] = ()


def _subjects_for_rewrite(
    action: RewriteAction,
    *,
    by_id: dict[int, Any],
    familiar_display_name: str | None,
) -> list[FactSubject]:
    """Build ``FactSubject`` list for a rewrite's chosen keys.

    display_at_write resolved from the matching source-fact subject
    when present; for the self key falls back to the familiar's display
    name; otherwise the key's local part. Keeps display resolution
    working at read time without a second lookup.
    """
    display_by_key: dict[str, str] = {}
    for fid in action.old_fact_ids:
        for s in by_id[fid].subjects:
            display_by_key.setdefault(s.canonical_key, s.display_at_write)
    out: list[FactSubject] = []
    for key in action.subject_keys:
        if key in display_by_key:
            display = display_by_key[key]
        elif is_self_key(key):
            display = familiar_display_name or key.split(":", 1)[-1]
        else:
            display = key.split(":", 1)[-1]
        out.append(FactSubject(canonical_key=key, display_at_write=display))
    return out


async def apply_consolidation(
    store: AsyncHistoryStore,
    plan: ConsolidationPlan,
    *,
    familiar_display_name: str | None = None,
) -> ApplyReport:
    """Execute a plan's accepted actions; advance the sleep watermark."""
    fam = plan.familiar_id

    # snapshot rewrite sources up front (exact id fetch, includes
    # superseded) to resolve subject displays + the merge channel even as
    # rows get superseded — and so a >100k fact base can't drop a needed
    # row. Provenance union is the store's job now (see supersede).
    needed: set[int] = set()
    for a in plan.rewrite:
        needed.update(a.old_fact_ids)
    by_id = {f.id: f for f in await store.facts_by_ids(familiar_id=fam, ids=needed)}

    skipped: list[tuple[str, int, str]] = []
    retired: list[int] = []
    for action in plan.retire:
        # the store retires each id, recording (not raising on) any a
        # concurrent writer already retired/superseded since plan time.
        result = await store.supersede(
            familiar_id=fam, obsolete_facts=action.fact_ids, new_fact=None
        )
        retired.extend(result.superseded)
        skipped.extend(("retire", fid, reason) for fid, reason in result.skipped)

    rewritten: list[tuple[tuple[int, ...], int]] = []
    for action in plan.rewrite:
        # the store owns the merge: it pre-flights every source, unions
        # their provenance, and mints (dedup=False) only if all are
        # current — declining the whole merge otherwise. Subjects stay
        # caller-prepared; channel follows the first obsolete row.
        result = await store.supersede(
            familiar_id=fam,
            obsolete_facts=action.old_fact_ids,
            new_fact=FactDraft(
                channel_id=by_id[action.old_fact_ids[0]].channel_id,
                text=action.new_text,
                subjects=tuple(
                    _subjects_for_rewrite(
                        action,
                        by_id=by_id,
                        familiar_display_name=familiar_display_name,
                    )
                ),
            ),
        )
        if result.minted is None:
            # a source raced between plan + apply, so the store declined
            # the merge whole — same outcome as the old pre-flight skip:
            # nothing minted, the action recorded as skipped, no raise.
            skipped.extend(("rewrite", fid, reason) for fid, reason in result.skipped)
            continue
        rewritten.append((action.old_fact_ids, result.minted.id))

    # consolidation owns the FACT axis only; the turn axis belongs to the
    # opinion pass. Advancing just last_fact_id leaves its progress untouched.
    await store.advance_sleep_watermark(
        familiar_id=fam,
        last_fact_id=plan.new_last_fact_id,
    )
    _logger.info(
        "sleep-consolidation applied familiar=%s retired=%d rewritten=%d "
        "skipped=%d fact_watermark=%d",
        fam,
        len(retired),
        len(rewritten),
        len(skipped),
        plan.new_last_fact_id,
    )
    return ApplyReport(
        retired_fact_ids=tuple(retired),
        rewritten=tuple(rewritten),
        watermark=(plan.new_last_fact_id, plan.new_last_turn_id),
        skipped=tuple(skipped),
    )

File: rust/familiar-connect/src/sleep/apply.rs
//! Apply a validated [`ConsolidationPlan`] to the DB (subsystem 04; Python
//! `sleep/apply.py`).
//!
//! Separate from planning so a dry run can compute the plan without touching the
//! DB. Apply is supersede-only: retires drop facts with no replacement, rewrites
//! append one consolidated fact then supersede the old ones by it. The sleep
//! watermark advances once, on the **fact axis only** (the turn axis belongs to
//! the opinion pass), regardless of how many actions ran.

use std::collections::HashMap;

use super::consolidation::{ConsolidationPlan, RewriteAction};
use crate::history::async_store::AsyncHistoryStore;
use crate::history::store::{Fact, FactDraft, FactSubject, NewFact};
use crate::identity::is_ego_key;

/// Outcome of applying a plan.
///
/// `skipped` records actions a concurrent writer invalidated between plan and
/// apply — each `(kind, fact_id, reason)`. Per-action skip-and-record makes
/// apply non-fatal and idempotent: a partial or already-applied plan re-runs
/// cleanly instead of crashing.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ApplyReport {
    /// Obsolete ids actually retired this call.
    pub retired_fact_ids: Vec<i64>,
    /// `(old_ids, new_id)` for each rewrite that minted a replacement.
    pub rewritten: Vec<(Vec<i64>, i64)>,
    /// `(last_fact_id, last_turn_id)` — the turn element is informational only
    /// (this pass writes only the fact axis).
    pub watermark: (i64, i64),
    /// `(kind, fact_id, reason)` for rows a concurrent writer invalidated.
    pub skipped: Vec<(String, i64, String)>,
}

/// Build the `FactSubject` list for a rewrite's chosen keys.
///
/// `display_at_write` resolves from the matching source-fact subject when
/// present (first-seen wins, iterating `old_fact_ids` order then subject order);
/// for an ego key it falls back to `familiar_display_name` or the key's local
/// part; otherwise the key's local part.
///
/// # Errors
/// Raises when a source id is absent from `by_id` — Python indexes `by_id[fid]`,
/// raising `KeyError`; supersede-only removal keeps every plan-time id fetchable
/// by exact id at apply time, so this is unreachable in the pipeline.
fn subjects_for_rewrite(
    action: &RewriteAction,
    by_id: &HashMap<i64, Fact>,
    familiar_display_name: Option<&str>,
) -> anyhow::Result<Vec<FactSubject>> {
    let mut display_by_key: HashMap<String, String> = HashMap::new();
    for fid in &action.old_fact_ids {
        let f = by_id
            .get(fid)
            .ok_or_else(|| anyhow::anyhow!("rewrite source fact {fid} missing from snapshot"))?;
        for s in &f.subjects {
            display_by_key
                .entry(s.canonical_key.clone())
                .or_insert_with(|| s.display_at_write.clone());
        }
    }
    let mut out: Vec<FactSubject> = Vec::new();
    for key in &action.subject_keys {
        let display = display_by_key.get(key).map_or_else(
            || {
                if is_ego_key(key) {
                    familiar_display_name.map_or_else(|| local_part(key), ToOwned::to_owned)
                } else {
                    local_part(key)
                }
            },
            Clone::clone,
        );
        out.push(FactSubject {
            canonical_key: key.clone(),
            display_at_write: display,
        });
    }
    Ok(out)
}

/// Python `key.split(":", 1)[-1]`: everything after the first `:`, or the whole
/// string when there is no colon.
fn local_part(key: &str) -> String {
    key.split_once(':')
        .map_or_else(|| key.to_owned(), |(_, rest)| rest.to_owned())
}

/// Execute a plan's accepted actions; advance the sleep watermark (fact axis).
///
/// This is the ONLY consolidation mutator — planning is dry-run-safe by
/// construction. Apply order: snapshot all rewrite source facts up front (exact
/// id fetch, includes superseded), execute all retires, then all rewrites, then
/// advance the fact-axis watermark once (even when everything skipped).
///
/// # Errors
/// Propagates store faults.
pub async fn apply_consolidation(
    store: &AsyncHistoryStore,
    plan: &ConsolidationPlan,
    familiar_display_name: Option<&str>,
) -> anyhow::Result<ApplyReport> {
    let fam = &plan.familiar_id;

    // snapshot rewrite sources up front (exact id fetch, includes superseded) to
    // resolve subject displays + the merge channel even as rows get superseded.
    let mut needed: Vec<i64> = Vec::new();
    for a in &plan.rewrite {
        for &id in &a.old_fact_ids {
            if !needed.contains(&id) {
                needed.push(id);
            }
        }
    }
    let by_id: HashMap<i64, Fact> = store
        .facts_by_ids(fam.clone(), needed)
        .await?
        .into_iter()
        .map(|f| (f.id, f))
        .collect();

    let mut skipped: Vec<(String, i64, String)> = Vec::new();
    let mut retired: Vec<i64> = Vec::new();
    for action in &plan.retire {
        // the store retires each id, recording (not raising on) any a concurrent
        // writer already retired/superseded since plan time.
        let result = store
            .supersede(fam.clone(), action.fact_ids.clone(), NewFact::Retire)
            .await?;
        retired.extend(result.superseded);
        skipped.extend(
            result
                .skipped
                .into_iter()
                .map(|(fid, reason)| ("retire".to_owned(), fid, reason)),
        );
    }

    let mut rewritten: Vec<(Vec<i64>, i64)> = Vec::new();
    for action in &plan.rewrite {
        // the store owns the merge: it pre-flights every source, unions their
        // provenance, and mints only if all are current — declining the whole
        // merge otherwise. Subjects stay caller-prepared; channel follows the
        // first obsolete row. A source id absent from the snapshot raises
        // (Python indexes `by_id[...]`, a KeyError) rather than degrading —
        // unreachable in the pipeline, where supersede is the only removal path.
        let channel_id = by_id
            .get(&action.old_fact_ids[0])
            .ok_or_else(|| {
                anyhow::anyhow!(
                    "rewrite source fact {} missing from snapshot",
                    action.old_fact_ids[0]
                )
            })?
            .channel_id;
        let draft = FactDraft {
            channel_id,
            text: action.new_text.clone(),
            subjects: subjects_for_rewrite(action, &by_id, familiar_display_name)?,
        };
        let result = store
            .supersede(
                fam.clone(),
                action.old_fact_ids.clone(),
                NewFact::Merge(draft),
            )
            .await?;
        match result.minted {
            Some(minted) => rewritten.push((action.old_fact_ids.clone(), minted.id)),
            None => {
                // a source raced between plan + apply — the store declined the
                // whole merge; record every skip and continue.
                skipped.extend(
                    result
                        .skipped
                        .into_iter()
                        .map(|(fid, reason)| ("rewrite".to_owned(), fid, reason)),
                );
            }
        }
    }

    // consolidation owns the FACT axis only; the turn axis belongs to the
    // opinion pass. Advancing just last_fact_id leaves its progress untouched.
    store
        .advance_sleep_watermark(fam.clone(), Some(plan.new_last_fact_id), None)
        .await?;
    tracing::info!(
        target: "familiar_connect.sleep.apply",
        "sleep-consolidation applied familiar={fam} retired={} rewritten={} skipped={} fact_watermark={}",
        retired.len(),
        rewritten.len(),
        skipped.len(),
        plan.new_last_fact_id,
    );
    Ok(ApplyReport {
        retired_fact_ids: retired,
        rewritten,
        watermark: (plan.new_last_fact_id, plan.new_last_turn_id),
        skipped,
    })
}
